Fix stretch histogram reset_limits

astrofrog opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This should fix reset_limits in the stretch histogram


Fixes #2506

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

pllim commented:
@kecnry , should this be backported? The code for this plugin in v3.7.x look very different.

pllim commented on jdaviz/configs/default/plugins/plot_options/plot_options.py:
```diff
@@ -640,7 +640,14 @@ def _update_stretch_histogram(self, msg={}):
+        # change the x_att to the dummy 'ref' dataset then change it back to
+        # the latest ComponentID after.
+
+        self.stretch_histogram.viewer.state.x_att = self.stretch_histogram.app.data_collection['ref'].id['x']  # noqa: E501
```
@astrofrog , do we need "hold sync" block here?

pllim commented:
Now the CI failure is real:

```
E           glue.core.exceptions.IncompatibleAttribute: x

glue/core/data.py:1467: IncompatibleAttribute
```

astrofrog commented:
I'll investigate the failure, it is indeed a real problem.

kecnry commented:
> should this be backported? 

No need to backport this, no.  The migration to using a glue viewer is unreleased.

astrofrog commented:
I found a bug deep inside glue that was causing issues here:

* https://github.com/glue-viz/glue/pull/2451

I've released glue-core v1.14.1 with a fix for that bug. I've now removed the workaround that I originally added in this PR and have also removed the workaround to add the 'ref' dataset at all as this appears to no longer be needed with the above bug fix.

One thing that is missing at the moment in a regression test for the original bug here - it might be better for @kecnry to add this as I'm not sure what the best way is to do this via public API.

astrofrog commented on CHANGES.rst:
```diff
@@ -56,6 +56,8 @@ Specviz2d
 Bug Fixes
 ---------
 
+- Fixed histogram viewer breaking when its "Home" button is pressed within Plot Options. [#2529]
```
So technically this isn't fixed by this PR but by glue-core v1.14.1, but maybe still worth leaving this here?

kecnry replied:
Since the bug itself never saw a release, maybe we should just lump this PR in with one of the other changelog entries?  Maybe alongside #2498?

astrofrog replied:
If the bug never saw a release we could just remove any mention of it altogether from the changelog?

astrofrog commented:
I removed the changelog entry and marked this as affects-dev

kecnry commented:
I'm still seeing some weird behavior.  Would you expect `reset_limits` to also affect `viewer.set.hist_x_min`/`hist_x_max`?  But even once I reset those to the originally set value, the bins still don't look the same as they did before resetting the limits 🤔

astrofrog commented:
@kecnry - I'll invedstigate!

astrofrog commented:
@kecnry - from what I can see, the stretch histogram in the plot options manually sets the histogram limits to be the 95% percentile limits of the data. However, ``reset_limits`` as implemented on the state object for the histogram viewer uses the strict min/max of the data to set the histogram limits. In addition, ``reset_limits`` does cause ``state.update_bins_to_view`` to be called, so when you call ``state.reset_limits``, it will set both ``x_min``/``x_max`` and ``hist_x_min``/``hist_x_max`` to be the strict min/max of the data.

So there are three questions here:

* Should we make it so that ``reset_limits`` in glue has the option to have a customizable percentile to use for the min/max? This would be easy enough to adjust in glue-core (see https://github.com/glue-viz/glue/blob/4de2cfd31670acb9aae60d198be4a5f83ebd8c0e/glue/viewers/histogram/state.py#L70) as we can simply set the percentile to a variable instead of 100%. This would then avoid jdaviz having to implement custom percentile interval logic and would make the home button work as you expect.

* Do you want reset_limits to *not* reset the bins being used, or would this be fine provided the bins were also reset to e.g. the 95% percentile limits if we implement the above?

* Do you think there is another issue going on not described by the above?

pllim commented:
The 95th percentile rings a bell. I think POs wanted the default to be 95th percentile or something because min/max was not displaying anything meaningful in real data.

kecnry commented:
That's a good point - we may want bypass the home button eventually to set the limits based on the current locations of vmin/vmax, but I think `reset_limits` still needs to be fixed either way.

astrofrog commented:
@kecnry - see https://github.com/glue-viz/glue/pull/2455, with this you can do:

```
po._obj.stretch_histogram.viewer.state.update_bins_on_reset_limits = False
po._obj.stretch_histogram.viewer.state.x_limits_percentile = 95
```

which I think would give the behavior you want. Could you test it out and let me know if this meets your needs? If so, I can merge & release.

kecnry commented:
@astrofrog - that seems to do the trick and act as expected thanks!  I added a commit here which will make use of that once its release and the pin for glue updated in this PR.
